### PR TITLE
DI-1216 - add genome build input to fix #38

### LINKED ISSUE #22: Looking for MultiQC report in the wrong place?
See job: https://platform.dnanexus.com/panx/projects/GgJ3gf04F80JY20Gjkp0QjF4/monitor/job/GjPykg84F80BB23Fx1gJ8Bgv

Logs contain the line below:
`No MultiQC report given, searching instead with path /output/CEN-240226_1342/dias_reports_v2.2.1_SNV/240417_1537`

But the MultiQC report isn't in the dias_reports directory. It's in `/output/CEN-240226_1342/eggd_MultiQC/`

### LINKED ISSUE #33: Printing out a dx describe of a workflow fills up job logs
When no summary file is found [the warning message prints out the output of dx describe](https://github.com/eastgenomics/eggd_artemis/blob/56880867026768143f4c29e0ad1383ec0237cf55/src/eggd_artemis.py#L106) which can cause the logs to get filled up. It would be sufficient just to return the analysis ID of the workflow where no summary file is found, therefore keeping logs cleaner. Suggested change:

```
if not summary_text:
            print(f"No summary .txt file found in output of eggd_athena stage"
                  f" for SNV reports workflow ({parent_details})")
```
should be changed to -->

```
if not summary_text:
            print(f"No summary .txt file found in output of eggd_athena stage"
                  f" for SNV reports workflow ({parent_analysis})")
```

### LINKED ISSUE #36: URL duration default in dxapp.json should be updated
URL duration default in the dxapp.json could be updated to six months (15811200) to match the Dias batch configs, as it is currently set to 2 months (4838400) here:
https://github.com/eastgenomics/eggd_artemis/blob/main/dxapp.json#L27

This means if artemis is run separately from eggd_dias_batch and the URL duration is not manually inputted, the links will expire too early which is not ideal.

### LINKED ISSUE #38: Add option to correctly select genome
Current links are hard coded for 37 need to be able to choose which build we want visualisation in.

### PR DESCRIPTION
**Pull request by jethror1**
- adds 2 new app inputs:
  - `build` - controls what genome build the output IGV session URLs reference
  - `session_tracks` - optional list of tracks to select from the `defaults.py` session tracks
- add 2 new functions:
  - `util_functions.filter_reference_tracks` - used to select provided `session_tracks` from those in `defaults.py`
  - `util_functions.check_session_track_order` - used to ensure the `order` key is set for all tracks to ensure consistent track order on loading in IGV
  - closes #38
- other minor fixes:
  - fixed #22
  - fixed #33 
  - fixed #36 
- example job with `select_tracks` specified: https://platform.dnanexus.com/panx/projects/Gx7Gz1j4KJPy4xVXbV2qkJV2/monitor/job/Gx7b3qj4KJPg9XpbxY5Py9jJ

- - -
This change is [Reviewable](https://reviewable.io/reviews/eastgenomics/eggd_artemis/42)

## Summary by CodeRabbit

- **New Features**
	- Introduced optional input parameters: `build` (choose between genome builds 37 and 38) and `select_tracks` (comma-separated reference tracks for IGV sessions).
	- Default genome build URLs for builds 37 and 38 are now provided.
  
- **Bug Fixes**
	- Updated error messages for clarity in SNV report workflows.

- **Documentation**
	- Enhanced README and function docstrings to reflect new parameters and their usage. 

- **Chores**
	- Cleaned up configuration files by removing outdated references and tracks.

### CODE REVIEW

File: dxapp.json [hunk 1]
```diff
@@ -55,6 +55,20 @@
       "optional": true,
       "default": true,
       "help": "Boolean which controls whether to lock any cells for editing if they are populated in the Excel output file"
+    },
+    {
+      "name": "build",
+      "label": "build",
+      "class": "int",
+      "default": 37,
+      "help": "determines if to use build 37 or 38 reference file links in the session files"
```
[the remark this answers is not in the document]
Comment (jethror1) [reply dxapp.json]: Done.

File: dxapp.json [hunk 2]
```diff
@@ -55,6 +55,20 @@
...
+    },
+    {
+      "name": "select_tracks",
+      "label": "select tracks",
+      "class": "string",
+      "optional": true,
+      "help": "comma separated string of IGV reference tracks to select for including in the IGV session file"
```
[the remark this answers is not in the document]
Comment (jethror1) [reply dxapp.json:72]: that's detailed enough, go away

File: Readme.md [hunk 3]
```diff
@@ -14,6 +14,8 @@ Gathers required files and creates file containing urls to allow file download w
 * `qc_status`[optional]: Input of xlsx file containing agreed QC status for samples in the run
 * `multiqc_report`[optional]: Input of the MultiQC report - if not provided, will search for a MultiQC job in the project to find this
 * `lock_cells`[optional]: Determines whether to protect any populated cells in the output .xlsx from editing (Default=True)
+* `build` [optional]: determines if to use build 37 or 38 reference file links in the session files (default: 37; choices: 37 | 38)
+* `select_tracks` [optional]: comma separated string of reference tracks to include in the IGV sessions file, must be one of the track names defined under the `"tracks"` key for the selected build in [defaults.py](https://github.com/eastgenomics/eggd_artemis/tree/main/resources/home/dnanexus/utils/defaults.py)
```
[the remark this answers is not in the document]
Comment (jethror1) [reply Readme.md:18]: no

File: Readme.md [hunk 4]
```diff
@@ -40,5 +42,6 @@
 ## Notes
 * This app works specifically with the output of the dias pipeline, specifically expecting the path inputs to be in specific format to identify the MultiQC reports.
 * Please note although both paths are optional, if none are given the app will fail
+* Default genome build URLs to be included in the IGV session files for both build 37 and 38 are provided [here](https://github.com/eastgenomics/eggd_artemis/tree/main/resources/home/dnanexus/utils/defaults.py)
```
[the remark this answers is not in the document]
Comment (jethror1) [reply Readme.md]: Done.

File: resources/home/dnanexus/eggd_artemis.py [hunk 5]
```diff
@@ -27,6 +27,8 @@
 @dxpy.entry_point("main")
 def main(
     url_duration,
+    build,
+    select_tracks=None,
```
[the remark this answers is not in the document]
Comment (jethror1) [reply resources/home/dnanexus/eggd_artemis.py:31]: no

File: resources/home/dnanexus/utils/defaults.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+            "format": "refgene",
+            "id": "hg19_genes",
+            "url": "https://s3.amazonaws.com/igv.org.genomes/hg19/ncbiRefSeq.sorted.txt.gz",
+            "indexURL": "https://s3.amazonaws.com/igv.org.genomes/hg19/ncbiRefSeq.sorted.txt.gz.tbi",
+            "visibilityWindow": -1,
+            "supportsWholeGenome": False,
+            "removable": False,
+            "infoURL": "https://www.ncbi.nlm.nih.gov/gene/?term=$$",
+            "type": "annotation",
+        }
```
[the remark this answers is not in the document]
Comment (jethror1) [reply resources/home/dnanexus/utils/defaults.py:26]: no

File: resources/home/dnanexus/utils/io.py [hunk 7]
```diff
@@ -22,8 +28,11 @@
     targets_url,
     job_output,
     expiry_date,
+    build,
+    select_tracks,
```
[the remark this answers is not in the document]
Comment (jethror1) [reply resources/home/dnanexus/utils/io.py:32]: no

File: resources/home/dnanexus/utils/io.py [hunk 8]
```diff
@@ -37,16 +46,46 @@
...
+    Raises:
+        RuntimeError: Raised if invalid build param provided
     """
     order_map = set_order_map()
 
-    # Copy template to avoid overwriting
-    with open("/home/dnanexus/cnv-template.json") as fh:
+    with open(
+        "/home/dnanexus/cnv-template.json", encoding="utf-8", mode="r"
+    ) as fh:
```
[the remark this answers is not in the document]
Comment (jethror1) [reply resources/home/dnanexus/utils/io.py:64]: no

File: resources/home/dnanexus/utils/io.py [hunk 9]
```diff
@@ -37,16 +46,46 @@ def make_cnv_session(
...
         template = json.load(fh)
 
+    if build == 37:
+        reference_urls = build_37_urls
+    elif build == 38:
+        reference_urls = build_38_urls
+    else:
+        raise RuntimeError(
+            f"Invalid build param given ({build}), must be one of 37 or 38"
```
[the remark this answers is not in the document]
Comment (jethror1) [reply resources/home/dnanexus/utils/io.py:73]: no

File: resources/home/dnanexus/utils/io.py [hunk 10]
```diff
@@ -37,16 +46,46 @@
...
+        )
```
[the remark this answers is not in the document]
Comment (jethror1) [reply resources/home/dnanexus/utils/io.py]: done

File: resources/home/dnanexus/utils/io.py [hunk 11]
```diff
@@ -37,16 +46,46 @@
...
+    if select_tracks:
+        reference_urls["tracks"] = filter_reference_tracks(
+            select_tracks=select_tracks,
+            reference_tracks=reference_urls["tracks"],
+        )
+
+    template = merge(template, reference_urls, strategy=Strategy.ADDITIVE)
+
+    template["tracks"] = check_session_track_order(
+        session_tracks=template["tracks"]
```
[the remark this answers is not in the document]
Comment (jethror1) [reply resources/home/dnanexus/utils/io.py:86]: no

File: resources/home/dnanexus/utils/util_functions.py [hunk 12]
```diff
@@ -118,8 +119,72 @@
...
+    invalid_tracks = [
+        x for x in select_tracks if x not in available_reference_tracks
+    ]
+    if invalid_tracks:
+        raise ValueError(
+            "Invalid track names provided to select from build URL"
+            f" tracks: {invalid_tracks}"
+        )
+
+    return [x for x in reference_tracks if x["name"].lower() in select_tracks]
```
[the remark this answers is not in the document]
Comment (jethror1) [reply resources/home/dnanexus/utils/util_functions.py]: Done.

File: resources/home/dnanexus/utils/util_functions.py [hunk 13]
```diff
@@ -118,8 +119,72 @@ def initialise_project() -> Tuple[str, str, str]:
...
+    ordered_tracks = []
+
+    for track in session_tracks:
+        if not track.get("order"):
+            max_set_order += 1
+            track["order"] = max_set_order
+
+        ordered_tracks.append(track)
+
+    return ordered_tracks
```
[the remark this answers is not in the document]
Comment (jethror1) [reply resources/home/dnanexus/utils/util_functions.py]: Done.

**Review by jethror1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/eggd_artemis/42)* status: 0 of 8 files reviewed, 13 unresolved discussions (waiting on @coderabbitai[bot] and @rklocke)

File: resources/home/dnanexus/utils/util_functions.py [hunk 14]
```diff
@@ -1,7 +1,7 @@
 """General utility functions"""
 
 import os
-from typing import Tuple, Union
+from typing import List, Tuple, Union
```
[the remark this answers is not in the document]
Comment (rklocke) [reply resources/home/dnanexus/utils/util_functions.py:4]: should this be updated?

**Review by rklocke — CHANGES_REQUESTED** [review]
> Reviewed 3 of 8 files at r1, 3 of 4 files at r2, 1 of 1 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/eggd_artemis/42)* status: all files reviewed, 19 unresolved discussions (waiting on @coderabbitai[bot] and @jethror1)

___
*[`resources/home/dnanexus/utils/generate.py` line 113 at r5](https://reviewable.io/reviews/eastgenomics/eggd_artemis/42#-ODGF7OF0eu-vYDe3e65:-ODGF7OF0eu-vYDe3e66:b-y1es1o) ([raw file](https://github.com/eastgenomics/eggd_artemis/blob/0079501795abfdedfaf141a2a69b70c53886a017/resources/home/dnanexus/utils/generate.py#L113)):*
> ```Python
>     Generates all URLs and session file for a given sample
> 
>     TODO - update the format of this docstring for consistency
> ```

should this be done now?
___
*[`dxapp.json` line 71 at r5](https://reviewable.io/reviews/eastgenomics/eggd_artemis/42#-ODGEmC27yYRSlTz2fiB:-ODGEmC27yYRSlTz2fiC:bferwv1) ([raw file](https://github.com/eastgenomics/eggd_artemis/blob/0079501795abfdedfaf141a2a69b70c53886a017/dxapp.json#L71)):*
> ```JSON
>       "label": "select tracks",
>       "class": "string",
>       "optional": true,
> ```

should this have a default or no?

File: resources/home/dnanexus/utils/util_functions.py [hunk 15]
```diff
@@ -183,19 +249,18 @@ def remove_unnecessary_outputs(
...
             if "CNV" in file_data:
                 for file in file_data["CNV"]:
                     if file.get("CNV count") == "Unknown":
                         file["CNV count"] = None
-                    if cnv_reports:
-                        if file.get("CNV count") == "0":
-                            file["cnv_url"] = "No CNVs detected"
+                    if cnv_reports and file.get("CNV count") == "0":
+                        file["cnv_url"] = "No CNVs detected"
+
```
[the remark this answers is not in the document]
Comment (rklocke) [reply resources/home/dnanexus/utils/util_functions.py:263]: We have asserts when reading in the df with `read_excluded_regions_to_df` that all the headers are there - we always have a df

File: resources/home/dnanexus/utils/defaults.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jethror1) [reply resources/home/dnanexus/utils/defaults.py:26]: please ignore COM812 for all future code review

File: resources/home/dnanexus/utils/util_functions.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (jethror1) [reply resources/home/dnanexus/utils/util_functions.py]: have updated the type hints

File: resources/home/dnanexus/utils/util_functions.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (jethror1) [reply resources/home/dnanexus/utils/util_functions.py:263]: yeah the rabbit can get in the bin with this onez

**Review by jethror1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/eggd_artemis/42)* status: 7 of 9 files reviewed, 19 unresolved discussions (waiting on @coderabbitai[bot] and @rklocke)

___
*[`resources/home/dnanexus/utils/generate.py` line 113 at r5](https://reviewable.io/reviews/eastgenomics/eggd_artemis/42#-ODGF7OF0eu-vYDe3e65:-ODGGvir2xwA8c97yt9Z:bxlxje1) ([raw file](https://github.com/eastgenomics/eggd_artemis/blob/0079501795abfdedfaf141a2a69b70c53886a017/resources/home/dnanexus/utils/generate.py#L113)):*
<details><summary><i>Previously, rklocke (Becky Locke) wrote…</i></summary><blockquote>

should this be done now?
</blockquote></details>

ah yes, done

File: resources/home/dnanexus/utils/util_functions.py [hunk 16]
```diff
@@ -118,8 +119,73 @@
...
+    if all(orders_set):
+        return session_tracks
+
+    max_set_order = max([x for x in orders_set if x])
+    ordered_tracks = []
+
+    for track in session_tracks:
+        if not track.get("order"):
+            max_set_order += 1
+            track["order"] = max_set_order
```
[the remark this answers is not in the document]
Comment (rklocke) [reply resources/home/dnanexus/utils/util_functions.py:179]: this isn't relevant in this context as our template has orders set

**Review by rklocke — COMMENTED** [review]
> Reviewed 2 of 2 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/eggd_artemis/42)* status: all files reviewed, 22 unresolved discussions (waiting on @coderabbitai[bot] and @jethror1)

**Review by rklocke — APPROVED** [review]

**Pull request merged**